README.md
# integration-kenshoo

[![CircleCI](https://circleci.com/gh/segment-integrations/integration-kenshoo.svg?style=shield&circle-token=33ad1cd091f09523e9342a118c3739778be914b9)](https://circleci.com/gh/segment-integrations/integration-kenshoo)
  
Kenshoo server-side integration for [Segment](https://segment.com).


HISTORY.md

2.0.3 / 2017-04-17
==================

  * Bump int-tester version to 2.x (#12)
2.0.0 / 2017-02-07
==================

  * removed versioned integration, back to v1

1.1.0 / 2017-01-31
==================

  * Standardize integration (linting, Docker configuration, circle.yml, upgrade
segmentio-integration version, upgrade integration-worker version, etc.)


test/index.js
/**
 * Created by moshee on 26/04/16.
 */
var Test = require('segmentio-integration-tester');
var helpers = require('./helpers');
var assert = require('assert');
var should = require('should');
var Kenshoo = require('..');
var mapper = require('../lib/mapper');
var facade = require('segmentio-facade');
var sinon = require('sinon');

describe("Kenshoo / Segment integration", function() {
  var settings, kenshoo, test, sandbox;

  beforeEach(function(){
    sandbox = sinon.sandbox.create();
    settings = {
      token: 'token',
      subdomain: '4075',
      eventToTypeMap: {
        'eventName Test': 'conv',
        'event name 1': 'click',
        'some-event': 'conv'
      }
    };
    kenshoo = new Kenshoo(settings);
    test = Test(kenshoo, __dirname);
  });

  it('should have the correct settings', function() {
    test
      .name('Kenshoo')
      .channels(['server', 'mobile', 'client'])
      .ensure('settings.token')
      .ensure('settings.subdomain')
  });

  describe('.validate()', function(){
    var msg;

    beforeEach(function(){
      msg = {
        event: 'some-event',
        context: {
          referrer: {
            id: 'clickId',
            type: 'kenshoo'
          }
        }
      };
    });

    it('should be invalid if `referrer.id` is missing', function(){
      delete msg.context.referrer.id;
      test.invalid(msg, settings);
    });

    it('should be invalid if `.referrer.type` is not "kenshoo"', function(){
      msg.context.referrer.type = 'foo';
      test.invalid(msg, settings);
    });

    it('should be valid if settings are complete', function(){
      test.valid(msg, settings);
    });
  });

  describe('verify event mapping to kenshoo payload', function() {
    var options = {};
    it('should map basic track with values', function() {
      var json = test.fixture('track-basic');
      var mapped = mapper.track(toMessage(json.input), settings, options);
      assert.deepEqual(mapped.payload, json.output);
    });

    it('should map basic track with no values and assign defaults', function() {
      var json = test.fixture('track-basic-no-defaults');
      var mapped = mapper.track(toMessage(json.input), settings, options);
      assert.deepEqual(mapped.payload, json.output);
    });

    it('should not return a payload for an unmapped event', function() {
      var json = test.fixture('track-basic-no-defaults');
      json.input.event = 'nope';
      var mapped = mapper.track(toMessage(json.input), settings);
      assert(mapped === undefined);
    });

    it('should extract the referrerId value', function() {
      var json = test.fixture('track-basic');
      var mapped = mapper.track(toMessage(json.input), settings, options);
      assert.equal(mapped.cookie, 'kenshoo_id=KENSHOO_CLICK_ID');
    });
  });

  function toMessage(msg){
    var msg = msg || {};
    return new facade.Track(msg);
  }

  describe('.track()', function(){
    var track;
    beforeEach(function() {
      track = helpers.track.bare({
        event: 'some-event',
        context: {
          referrer: {
            id: 'SAhCht3bQKPGd-z7ZvtYWW-i',
            type: 'kenshoo'
          }
        }
      });
    });

    it('should track an event successfully', function(done){
      test
        .track(track, settings)
        .expects(200, done);
    });

    it('should match the event name case-insensitively', function(done){
      var trackData = track.json();
      trackData.event = track.event().toUpperCase();
      track = helpers.track(trackData);
      test
        .track(track, settings)
        .expects(200, done);
    });

    it('should skip unmapped names', function(done){
      var spy = sandbox.spy(kenshoo, 'get');
      var trackData = track.json();
      trackData.event = 'nope';
      track = helpers.track(trackData);
      test
        .track(track, settings)
        .end(function(err, res){
          assert(!spy.called, 'Expected spy to have not been called');
          done(err);
        });
    });
  });
});


test/helpers.js
/**
 * Created by moshee
 * on: 26/04/16.
 */

var facade = require('segmentio-facade');
var extend = require('extend');
var uid = require('uid');

var userId1  = uid();
var email    = 'testing-' + userId1 + '@segment.io';

exports.track = function (context) {
    context = extend({
        userId     : userId1,
        event      : 'Baked a cake',
        properties : {
            revenue : 19.95,
            date : (new Date()).toISOString()
        },
        channel    : 'server',
        timestamp  : new Date(),
        context : {
            ip : '12.212.12.49',
            userAgent: 'Mozilla/5.0 (compatible; MSIE 10.0; Windows NT 6.2; Trident/6.0)'
        }
    }, context);
    return new facade.Track(context);
};


exports.track.bare = function (context) {
    context = extend({
        userId  : 'aaa',
        event   : 'Bear tracks',
        channel : 'server'
    }, context);
    return new facade.Track(context);
};


lib/index.js
/**
 * Created by moshee @ Kenshoo
 * on: 24/04/16.
 */

var Integration = require('segmentio-integration');
var mapper = require('./mapper');
var fmt = require('util').format;

/**
 * Kenshoo Segment integration module
 */

var Kenshoo = module.exports = new Integration('Kenshoo')
  .channels(['mobile', 'server', 'client'])
  .ensure('settings.token')
  .ensure('settings.subdomain')
  .ensure('settings.eventToTypeMap');

Kenshoo.ensure(function(msg) {
  var referrer = msg.proxy('context.referrer.type');
  if (referrer === 'kenshoo' || referrer === 'Kenshoo') return;
  return this.reject('referrer type must be Kenshoo');
})

Kenshoo.ensure(function(msg, settings) {
  if (msg.proxy('context.referrer.id')) return;
  return this.reject('kenshoo id required');
});

/**
 * Initialize.
 */

Kenshoo.prototype.initialize = function() {
  this.endpoint = fmt('https://%s.xg4ken.com/trk/v1', this.settings.subdomain);
};

/**
 * Track
 */

Kenshoo.prototype.track = function(track, fn) {
  var result = mapper.track(track, this.settings) || {};
  if (result.cookie && result.payload) {
    this.get()
      .set('User-Agent', 'Segment.io/1.0.0')
      .set('Cookie', result.cookie)
      .query(result.payload)
      .end(this.handle(fn));
  } else {
    this.debug('unmapped event');
    fn();
  }
};